src/tasks/tasks.hpp

```diff
@@ -0,0 +1,504 @@
+
+namespace parthenon {
+
+enum class TaskListStatus { complete }; // doesn't feel like we need this...
+enum class TaskType { normal, completion };
+
+class TaskQualifier {
+ public:
+  using qualifier_t = uint64_t;
+  TaskQualifier() = delete;
+  TaskQualifier(const qualifier_t n) : flags(n) {} // NOLINT(runtime/explicit)
+
+  static inline constexpr qualifier_t normal{0};
+  static inline constexpr qualifier_t local_sync{1 << 0};
+  static inline constexpr qualifier_t global_sync{1 << 1};
+  static inline constexpr qualifier_t completion{1 << 2};
+  static inline constexpr qualifier_t once_per_region{1 << 3};
+
+  bool LocalSync() const { return flags & local_sync; }
+  bool GlobalSync() const { return flags & global_sync; }
+  bool Completion() const { return flags & completion; }
+  bool Once() const { return flags & once_per_region; }
+  bool Valid() const { return true; }
+
+ private:
+  qualifier_t flags;
+};
+
+// forward declare Task for TaskID
+class Task;
+class TaskID {
+ public:
+  TaskID() : task(nullptr) {}
+  // pointers to Task are implicitly convertible to TaskID
+  TaskID(Task *t) : task(t) {} // NOLINT(runtime/explicit)
+
+  TaskID operator|(const TaskID &other) const {
+    // calling this operator means you're building a TaskID to hold a dependency
+    TaskID result;
+    if (task != nullptr)
+      result.dep.push_back(task);
+    else
+      result.dep.insert(result.dep.end(), dep.begin(), dep.end());
+    if (other.task != nullptr)
+      result.dep.push_back(other.task);
+    else
+      result.dep.insert(result.dep.end(), other.dep.begin(), other.dep.end());
+    return result;
+  }
+
+  const std::vector<Task *> &GetIDs() const { return std::cref(dep); }
+
+  bool empty() const { return (!task && dep.size() == 0); }
+  Task *GetTask() { return task; }
+
+ private:
+  Task *task = nullptr;
+  std::vector<Task *> dep;
+};
+
+class Task {
+ public:
+  Task() = default;
+  template <typename TID>
+  Task(TID &&dep, const std::function<TaskStatus()> &func,
+       std::pair<int, int> limits = {1, 1})
+      : f(func), exec_limits(limits) {
+    if (dep.GetIDs().size() == 0 && dep.GetTask()) {
+      dependencies.insert(dep.GetTask());
+    } else {
+      for (auto &d : dep.GetIDs()) {
+        dependencies.insert(d);
+      }
+    }
+    // always add "this" to repeat task if it's incomplete
+    dependent[static_cast<int>(TaskStatus::incomplete)].push_back(this);
+  }
+
+  TaskStatus operator()() {
+    auto status = f();
+    if (task_type == TaskType::completion) {
+      // keep track of how many times it's been called
+      num_calls += (status == TaskStatus::iterate || status == TaskStatus::complete);
+      // enforce minimum number of iterations
+      if (num_calls < exec_limits.first && status == TaskStatus::complete)
+        status = TaskStatus::iterate;
+      // enforce maximum number of iterations
+      if (num_calls == exec_limits.second) status = TaskStatus::complete;
```

Comment:
probably.  I suppose that would allow adding some kind of option that would warn or abort if an iteration failed to properly complete.  That seems like a good idea.  I'll either add an issue so I can deal with this later or add the feature if it looks simple enough.